### PR TITLE
Rl  stockdb

### PR DESCRIPTION
**Pull request by ralles-liu**
This adds the ability to pull stock information from Piquette Finance-go in our Stockings package

### CODE REVIEW

File: cmd/coattails/main.go [hunk 1]
```diff
@@ -4,18 +4,19 @@ import (
...
+	"time"
+
 	"github.com/bluedresscapital/coattails/pkg/routes"
 	"github.com/bluedresscapital/coattails/pkg/stockings"
 	"github.com/bluedresscapital/coattails/pkg/sundress"
 	"github.com/bluedresscapital/coattails/pkg/wardrobe"
 	"github.com/gorilla/mux"
 	"github.com/joho/godotenv"
 	"github.com/rs/cors"
-	"log"
```
Comment (jaysunmah) [inline cmd/coattails/main.go:14]: 🤔 is this in accordance with `go fmt`?

File: pkg/middleout/dbchecker.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package middleout
```
Comment (jaysunmah) [inline pkg/middleout/dbchecker.go]: lol what is this

File: pkg/middleout/dbchecker.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+// // assume there exists another file that handles the user requests
+// // that file gets the ticker and range of dates and will call the functions in here to interact with the database
+
+// // package just always seems to be declared as the director
+// package middleout	// could I declare something to be a package from a different director?
+
+// import (
```
Comment (Rsdutta) [inline pkg/middleout/dbchecker.go:9]: remove unused code

Comment (rsdutta) [inline pkg/middleout/dbchecker.go]: remove unused code

File: pkg/routes/fingotest.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
+package routes
+
+// import (
```
Comment (rsdutta) [inline pkg/routes/fingotest.go]: same here remove unused code

File: pkg/routes/auth.go [hunk 5]
```diff
@@ -96,6 +97,7 @@ func loginRegisterHelper(w http.ResponseWriter, r *http.Request, loginMode bool)
 	} else {
 		tok, err = auth.Register(l.Username, cipherPwd)
 		if err != nil {
+			log.Printf("we boomed here: %v", err)
```
Comment (jaysunmah) [inline pkg/routes/auth.go]: can u remove this hehe

File: pkg/routes/apitest.go [hunk 6]
```diff
@@ -11,8 +11,8 @@ import (
 func apiCheck(w http.ResponseWriter, r *http.Request) {
 	api := stockings.IexApi{}
 	aObj, err := api.GetCurrentPrice("MELI")
-	a, _ := json.Marshal(aObj)
-	fmt.Fprintln(w, string(a))
+	//a, _ := json.Marshal(aObj)
```
Comment (jaysunmah) [inline pkg/routes/apitest.go]: is this change intentional?

File: pkg/stockings/iex.go [hunk 7]
```diff
@@ -4,9 +4,10 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
-	"github.com/shopspring/decimal"
 	"net/http"
 	"time"
+
+	"github.com/shopspring/decimal"
```
Comment (jaysunmah) [inline pkg/stockings/iex.go:10]: is this go fmt'd as well?

File: pkg/stockings/stock.go [hunk 8]
```diff
@@ -18,6 +19,10 @@ type StockAPI interface {
 	GetHistoricalRange(ticker string, start time.Time, end time.Time) (*HistoricalStocks, error)
 }
 
+const (
+	DateLayout = "20060102"
```
Comment (jaysunmah) [inline pkg/stockings/stock.go]: this date format is only applicable to iex api, so can you put it under `iex.go`?

File: pkg/stockings/stock.go [hunk 9]
```diff
@@ -1,8 +1,9 @@
 package stockings
 
 import (
-	"github.com/shopspring/decimal"
 	"time"
+
+	"github.com/shopspring/decimal"
```
Comment (jaysunmah) [inline pkg/stockings/stock.go:6]: doesn't seem like a gofmt format either

File: pkg/stockings/fingopack.go [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+	// converting into our accepted Fingo form
+	endDate := datetime.Datetime{
+		Year:  year,
+		Month: month,
+		Day:   day,
+	}
+
+	// this brings our month down by one
+	if month == 1 {
```
Comment (jaysunmah) [inline pkg/stockings/fingopack.go]: insane tech LOL but we should probably do something with golang's builtins (see my comment above)

File: pkg/stockings/fingopack.go [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+
+	quoteAtDate := new(finance.ChartBar)
+
+	// converting the time.Time struct into usuable integers to be fed into Fingo
+	year := date.Year()
+	month := int(date.Month())
+	day := date.Day()
+
+	// converting into our accepted Fingo form
+	endDate := datetime.Datetime{
```
Comment (jaysunmah) [inline pkg/stockings/fingopack.go]: so I looked into `github.com/piquette/finance-go/datetime`, they have a `datetime.New(*time.Time)` function which should be useful for us. golang's `time` package is really nice, and it has builtin time delta functions for you. So instead of having to specifically check for stuff like whether month is 1, you can do something like:
```
endDate := datetime.New(date)
start := date.Sub(date.AddDate(0, -1, 0))
startDate := datetime.New(start)
```
```

Comment (jaysunmah) [reply pkg/stockings/fingopack.go]: also, do we actually have to have a window of a month? Is that their smallest window?

File: pkg/stockings/fingopack.go [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+// GetHistoricalRange is nearly identical to GetHistoricalPrice, but there are some nuances - might not be able to create a seperate helper function
+func (piq FingoPack) GetHistoricalRange(ticker string, start time.Time, end time.Time) (*HistoricalStocks, error) {
+
+	// need to create a piqHistoricalStocks type that is a list of finance.ChartBar
+	historicalRange := new(piqHistoricalStocks)
+
+	// just setting our start and end date based on what we were given
+	startDate := datetime.Datetime{
```
Comment (jaysunmah) [inline pkg/stockings/fingopack.go]: same here, you can just use their `datetime.New()` function

File: pkg/stockings/fingopack.go [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+	priceChange := float32(quote.RegularMarketChange)
+	percentChange := float32(quote.RegularMarketChangePercent)
+
+	return &Stock{symbol, name, currentPrice, priceChange, percentChange}, nil
+
+}
+
+// GetHistoricalPrice is hacky because Fingo only has a function that returns quotes in a time span
+// This function takes in the desired date as the end date, then sets the start date as one month prior
+func (piq FingoPack) GetHistoricalPrice(ticker string, date time.Time) (*HistoricalStock, error) {
```
Comment (jaysunmah) [inline pkg/stockings/fingopack.go:38]: also, do you realize that `GetHistoricalPrice` is essentially doing what `GetHistoricalRange` is doing? :) can we simplify the two functions to share like 90% of the same code?

**Review by jaysunmah — CHANGES_REQUESTED** [review]
left a ton of comments. for the love of god nobody merge this LOL

File: pkg/stockings/stock.go (hunk 8 above)
Comment (Rsdutta) [reply pkg/stockings/stock.go:23]: this is used in a global scope so I thought it might as well be in stock.go

File: pkg/middleout/dbchecker.go (hunk 3 above)
Comment (ralles-liu) [reply pkg/middleout/dbchecker.go]: removed.

File: pkg/routes/fingotest.go (hunk 4 above)
Comment (ralles-liu) [reply pkg/routes/fingotest.go]: removed.

File: cmd/coattails/main.go (hunk 1 above)
Comment (ralles-liu) [reply cmd/coattails/main.go:14]: not to sure why my go fmt did this, i can manually move it back if you want

File: pkg/middleout/dbchecker.go (hunk 2 above)
Comment (ralles-liu) [reply pkg/middleout/dbchecker.go]: removed.

File: pkg/routes/auth.go (hunk 5 above)
Comment (ralles-liu) [reply pkg/routes/auth.go]: removed.

File: pkg/routes/apitest.go (hunk 6 above)
Comment (ralles-liu) [reply pkg/routes/apitest.go]: updated so that it properly prints current price on the website

File: pkg/stockings/iex.go (hunk 7 above)
Comment (ralles-liu) [reply pkg/stockings/iex.go:10]: think so, not too sure whats wrong with my  go fmt

File: pkg/stockings/stock.go (hunk 9 above)
Comment (ralles-liu) [reply pkg/stockings/stock.go:6]: will run go fmt in git as well, maybe rishov should just go fmt it too

File: pkg/stockings/stock.go (hunk 8 above)
Comment (ralles-liu) [reply pkg/stockings/stock.go]: done, just noting that fingo also uses this date format

File: pkg/stockings/fingopack.go (hunk 11 above)
Comment (ralles-liu) [reply pkg/stockings/fingopack.go]: updated based on your comments and reduced window to 5 days. giving ourselves some cushioning in case there's some apocalyptic scenario

File: pkg/stockings/fingopack.go (hunk 10 above)
Comment (ralles-liu) [reply pkg/stockings/fingopack.go]: used the AddDate function you mentioned below

File: pkg/stockings/fingopack.go (hunk 12 above)
Comment (ralles-liu) [reply pkg/stockings/fingopack.go]: done

File: pkg/stockings/fingopack.go (hunk 13 above)
Comment (ralles-liu) [reply pkg/stockings/fingopack.go:38]: created a helper function that takes in the ticker and dates and returns the iterator

File: pkg/routes/auth.go (hunk 5 above)
Comment (rsdutta) [reply pkg/routes/auth.go]: was this removed?

Comment (jaysunmah) [reply pkg/routes/auth.go]: LMAO

File: pkg/routes/routes.go [hunk 14]
```diff
@@ -37,6 +38,8 @@ func RegisterAllRoutes(r *mux.Router) {
 	r.HandleFunc("/test", testHandler)
 	r.HandleFunc("/websocket", testWebSocket)
 	r.HandleFunc("/apitest", apiCheck)
+	//r.HandleFunc("/middleout", middleOutTest)
```
Comment (jaysunmah) [inline pkg/routes/routes.go:41]: remove this

File: pkg/stockings/fingopack.go [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+// GetHistoricalPrice is hacky because Fingo only has a function that returns quotes in a time span
+// This function takes in the desired date as the end date, then sets the start date as five days prior
+func (piq FingoPack) GetHistoricalPrice(ticker string, date time.Time) (*HistoricalStock, error) {
+
+	quoteAtDate := new(finance.ChartBar)
+
+	// creating a hypothetical start five days before (is this long enough)
+	adjStart := date.AddDate(0, 0, -5)
```
Comment (jaysunmah) [inline pkg/stockings/fingopack.go]: 🤔 why are we arbitrarily subtracting 5 days now? seems like there's no min requirement for a date range, so why not just +1 and -1 to our date? Better to fail later and understand why than just arbitrarily save unnecessary date ranges

File: pkg/stockings/fingopack.go (hunk 13 above)
Comment (jaysunmah) [reply pkg/stockings/fingopack.go:38]: lol do you still not see for `GetHistoricalPrice` you can just literally call `GetHistoricalRange` where you feed in `ticker`, `date` (as endDate), and `date.AddDate(0, -1, 0)` (as startDate) ? like everything below this line can get cleaned up, and your `GetHistoricalPrice` is like two lines

File: pkg/stockings/stock.go (hunk 9 above)
Comment (jaysunmah) [reply pkg/stockings/stock.go:6]: we're gonna need to discord over this to see how your IDE is running go fmt

**Review by jaysunmah — CHANGES_REQUESTED** [review]
almost there :)

File: pkg/routes/auth.go (hunk 5 above)
Comment (ralles-liu) [reply pkg/routes/auth.go]: okay now removed i think

File: pkg/stockings/fingopack.go [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+func (piq FingoPack) GetHistoricalPrice(ticker string, date time.Time) (*HistoricalStock, error) {
+
+	api := FingoPack{}
+	historicalQuotes, err := api.GetHistoricalRange(ticker, date.AddDate(0, 0, -5), date)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(*historicalQuotes) < 0 {
```
Comment (jaysunmah) [inline pkg/stockings/fingopack.go:46]: good check :)

File: pkg/stockings/fingopack.go [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+
+	return &Stock{symbol, name, currentPrice, priceChange, percentChange}, nil
+
+}
+
+// we use the GetHistoricalRange function and set our start date to 5 days prior becuase there has to be a valid open market date in the range
+func (piq FingoPack) GetHistoricalPrice(ticker string, date time.Time) (*HistoricalStock, error) {
+
+	api := FingoPack{}
+	historicalQuotes, err := api.GetHistoricalRange(ticker, date.AddDate(0, 0, -5), date)
```
Comment (jaysunmah) [inline pkg/stockings/fingopack.go]: bro u just call `piq.GetHistoricalRange` (`piq` from line 38 is basically `self` in python)

**Review by jaysunmah — CHANGES_REQUESTED** [review]
lololol JUST ONE MORE STEP

**Review by jaysunmah — APPROVED** [review]
